### PR TITLE
Cleanup RegionIterator to use RocksDB features directly

### PR DESCRIPTION
**Pull request by breeswish**
###  What have you changed?

By utilizing some recent RocksDB features, like `seek_for_prev` and iterator bound, we can simplify our `RegionIterator`. Also moved some error cases to standalone functions to reduce the hot-path code size.

###  What is the type of the changes?

- Improvement (a change which is an improvement to an existing feature)

###  How is the PR tested?

- Unit test
- Integration test

### CODE REVIEW

**Comment by mahjonp** [comment]
Oops, there was a dns resolve error when upload result, the final result was pasted here:
```diff
@@                               Benchmark Diff                               @@
================================================================================
tidb: 518692c14cab1d901f92a773488fc288cea2d8ce
--- tikv: 1d4e1046db90eacce4857c563a535fd8514fa36e
+++ tikv: 6f2f4ac108a13c56d33936026c149f3a6359d41e
pd: d8ea9951ab51d0e0301f794118259cb4941033cb
================================================================================
01.sql duration: 17886.58 ms ± 2.98% (std=227.73) delta: -2.01% (p=0.071)
02.sql duration: 6902.03 ms ± 2.83% (std=124.58) delta: -1.09% (p=0.389)
03.sql duration: 16851.16 ms ± 1.08% (std=104.16) delta: -5.09% (p=0.000)
04.sql duration: 7094.92 ms ± 1.84% (std=83.36) delta: -4.10% (p=0.026)
06.sql duration: 9412.94 ms ± 0.33% (std=15.67) delta: -4.74% (p=0.005)
07.sql duration: 13771.95 ms ± 0.87% (std=82.21) delta: -2.18% (p=0.000)
08.sql duration: 12039.05 ms ± 2.85% (std=168.33) delta: -1.96% (p=0.005)
09.sql duration: 33042.18 ms ± 2.29% (std=436.20) delta: 1.42% (p=0.044)
10.sql duration: 9796.48 ms ± 2.07% (std=108.02) delta: -2.51% (p=0.006)
11.sql duration: 3480.00 ms ± 5.99% (std=124.26) delta: -1.09% (p=0.801)
12.sql duration: 11433.96 ms ± 6.91% (std=344.22) delta: -6.81% (p=0.095)
13.sql duration: 9230.13 ms ± 2.58% (std=132.93) delta: -0.96% (p=0.045)
14.sql duration: 10812.88 ms ± 2.45% (std=149.28) delta: -6.16% (p=0.047)
15.sql duration: 19445.34 ms ± 4.37% (std=406.84) delta: -2.84% (p=0.074)
16.sql duration: 3565.05 ms ± 10.61% (std=274.67) delta: -6.04% (p=0.287)
17.sql duration: 34184.97 ms ± 1.42% (std=269.18) delta: -0.82% (p=0.748)
18.sql duration: 54432.80 ms ± 4.74% (std=1402.07) delta: -3.59% (p=0.014)
19.sql duration: 14777.47 ms ± 1.15% (std=100.45) delta: -5.79% (p=0.002)
20.sql duration: 10926.50 ms ± 3.21% (std=215.21) delta: -0.41% (p=0.674)
21.sql duration: 28281.48 ms…
```

File: src/storage/kv/cursor.rs [hunk 1]
```diff
@@ -346,30 +348,34 @@ impl<I: Iterator> Cursor<I> {
...
+
+    #[inline(never)]
+    fn handle_error_status(&self, e: Error) -> Result<()> {
+        // Split out the error case to reduce hot-path code size.
+        CRITICAL_ERROR.with_label_values(&["rocksdb iter"]).inc();
+        if panic_when_unexpected_key_or_data() {
+            set_panic_mark();
+            panic!(
+                "failed to iterate: {:?}, min_key: {:?}, max_key: {:?}",
+                e, self.min_key, self.max_key,
```
Comment (youjiali1995) [inline src/storage/kv/cursor.rs]: Does it need `hex::encode_upper()`?

File: src/storage/kv/cursor.rs [hunk 2]
```diff
@@ -346,30 +348,34 @@ impl<I: Iterator> Cursor<I> {
...
+            );
+        } else {
+            error!(
+                "failed to iterate";
+                "min_key" => ?self.min_key,
```
Comment (youjiali1995) [inline src/storage/kv/cursor.rs]: ditto

File: src/raftstore/store/region_snapshot.rs [hunk 3]
```diff
@@ -262,111 +240,63 @@ impl RegionIterator {
...
-        if self.valid {
-            let key = self.iter.key();
-            self.valid = if forward {
-                key < self.end_key.as_slice()
-            } else {
-                key >= self.start_key.as_slice()
-            };
-        }
-        self.valid
+        crate::storage::Iterator::seek_to_first(&mut self.iter)
```
Comment (youjiali1995) [inline src/raftstore/store/region_snapshot.rs]: Why not implement `crate::storage::Iterator` directly?

**Comment by sticnarf** [comment]
Seems that avoid inlining has nothing to do with `RegionIterator` simplification and can be done in another pr  :)

**Pull request merged**